[FX] Test native callables in FX lowering

jamesr66a opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#43426 [FX] Test native callables in FX lowering**

This test exercises the case where we use FX to translate from Python
code to some native callable object

For the purposes of testing, we use ElementwiseInterpreter defined
in test_custom_class.cpp.

We test that we can
1) Construct a native callable from FX IR
2) Construct a drop-in replacement module that delegates to the
native callable rather than the original code
3) Run both the original code and native callable wrapper with
equivalent results
4) TorchScript compile the native callable wrapper and confirm
equivalent execution results with the reference
5) TorchScript serialize and deserialize the native callable
and confirm equivalent execution results with the reference

Differential Revision: [D23273427](https://our.internmc.facebook.com/intern/diff/D23273427)

Commit:
[FX] Native callables in FX lowering

[ghstack-poisoned]

Commit:
Update on "[FX] Test native callables in FX lowering"


This test exercises the case where we use FX to translate from Python
code to some native callable object

For the purposes of testing, we use ElementwiseInterpreter defined
in test_custom_class.cpp.

We test that we can
1) Construct a native callable from FX IR
2) Construct a drop-in replacement module that delegates to the
native callable rather than the original code
3) Run both the original code and native callable wrapper with
equivalent results
4) TorchScript compile the native callable wrapper and confirm
equivalent execution results with the reference
5) TorchScript serialize and deserialize the native callable
and confirm equivalent execution results with the reference

Differential Revision: [D23273427](https://our.internmc.facebook.com/intern/diff/D23273427)

[ghstack-poisoned]

zdevito commented on test/test_fx.py:
```diff
@@ -186,6 +189,142 @@ def test_resnet(self):
+
+            # Create a graph that: 1) Takes function arguments 2) Invokes the interpreter
+            # 3) Returns the speficied return value
+
```
`symbolic_trace(module, <delegate_that_considers Wrapper a root module>)` would do the same with a lot less code.

jamesr66a replied:
Yeah that would be ideal, but I can't get `__call__` exposed on TorchBind classes without breaking `hasattr`. I can add a TODO

zdevito commented on test/test_fx.py:
```diff
@@ -186,6 +189,142 @@ def test_resnet(self):
+                elif n.op == 'call_function':
+                    assert target in target_to_name, "Unsupported call target " + target
+                    arg_names = []
+                    for arg in args:
```
If this is a demo of how to use fx, then this should probably call `map_arg` eventhough it really wouldn't support anything that does. Maybe it should also make it more clear that it is not supporting certain features like kwargs.

Commit:
Update on "[FX] Test native callables in FX lowering"


This test exercises the case where we use FX to translate from Python
code to some native callable object

For the purposes of testing, we use ElementwiseInterpreter defined
in test_custom_class.cpp.

We test that we can
1) Construct a native callable from FX IR
2) Construct a drop-in replacement module that delegates to the
native callable rather than the original code
3) Run both the original code and native callable wrapper with
equivalent results
4) TorchScript compile the native callable wrapper and confirm
equivalent execution results with the reference
5) TorchScript serialize and deserialize the native callable
and confirm equivalent execution results with the reference

Differential Revision: [D23273427](https://our.internmc.facebook.com/intern/diff/D23273427)

[ghstack-poisoned]

Commit:
Update on "[FX] Test native callables in FX lowering"


This test exercises the case where we use FX to translate from Python
code to some native callable object

For the purposes of testing, we use ElementwiseInterpreter defined
in test_custom_class.cpp.

We test that we can
1) Construct a native callable from FX IR
2) Construct a drop-in replacement module that delegates to the
native callable rather than the original code
3) Run both the original code and native callable wrapper with
equivalent results
4) TorchScript compile the native callable wrapper and confirm
equivalent execution results with the reference
5) TorchScript serialize and deserialize the native callable
and confirm equivalent execution results with the reference

Differential Revision: [D23273427](https://our.internmc.facebook.com/intern/diff/D23273427)

[ghstack-poisoned]

Commit:
Update on "[FX] Test native callables in FX lowering"


This test exercises the case where we use FX to translate from Python
code to some native callable object

For the purposes of testing, we use ElementwiseInterpreter defined
in test_custom_class.cpp.

We test that we can
1) Construct a native callable from FX IR
2) Construct a drop-in replacement module that delegates to the
native callable rather than the original code
3) Run both the original code and native callable wrapper with
equivalent results
4) TorchScript compile the native callable wrapper and confirm
equivalent execution results with the reference
5) TorchScript serialize and deserialize the native callable
and confirm equivalent execution results with the reference

Differential Revision: [D23273427](https://our.internmc.facebook.com/intern/diff/D23273427)

[ghstack-poisoned]

Commit:
Update on "[FX] Test native callables in FX lowering"


This test exercises the case where we use FX to translate from Python
code to some native callable object

For the purposes of testing, we use ElementwiseInterpreter defined
in test_custom_class.cpp.

We test that we can
1) Construct a native callable from FX IR
2) Construct a drop-in replacement module that delegates to the
native callable rather than the original code
3) Run both the original code and native callable wrapper with
equivalent results
4) TorchScript compile the native callable wrapper and confirm
equivalent execution results with the reference
5) TorchScript serialize and deserialize the native callable
and confirm equivalent execution results with the reference

Differential Revision: [D23273427](https://our.internmc.facebook.com/intern/diff/D23273427)

[ghstack-poisoned]
